torchvision/models/quantization/googlenet.py

```diff
@@ -0,0 +1,174 @@
+_QuantizableGoogLeNetOutputs = QuantizableGoogLeNetOutputs
+
+
+def googlenet(pretrained_float_model=False, progress=True, **kwargs):
+    r"""GoogLeNet (Inception v1) model architecture from
+    `"Going Deeper with Convolutions" <http://arxiv.org/abs/1409.4842>`_.
+
+    Args:
+        pretrained_float_model (bool): If True, returns a model pre-trained on ImageNet
+        progress (bool): If True, displays a progress bar of the download to stderr
+        aux_logits (bool): If True, adds two auxiliary branches that can improve training.
+            Default: *False* when pretrained_float_model is True otherwise *True*
+        transform_input (bool): If True, preprocesses the input according to the method with which it
+            was trained on ImageNet. Default: *False*
+    """
+    if pretrained_float_model:
+        if 'transform_input' not in kwargs:
+            kwargs['transform_input'] = True
+        if 'aux_logits' not in kwargs:
+            kwargs['aux_logits'] = False
+        if kwargs['aux_logits']:
+            warnings.warn('auxiliary heads in the pretrained_float_model googlenet model are NOT pretrained_float_model, '
+                          'so make sure to train them')
+        original_aux_logits = kwargs['aux_logits']
+        kwargs['aux_logits'] = True
+        kwargs['init_weights'] = False
+        model = QuantizableGoogLeNet(**kwargs)
+        state_dict = load_state_dict_from_url(model_urls['googlenet'],
+                                              progress=progress)
+        model.load_state_dict(state_dict)
+        if not original_aux_logits:
+            model.aux_logits = False
+            del model.aux1, model.aux2
+        return model
+
+    return QuantizableGoogLeNet(**kwargs)
+
+
+class QuantizableBasicConv2d(googlenet_module.BasicConv2d):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.relu = nn.ReLU()
+
+    def forward(self, x):
+        x = self.conv(x)
+        x = self.bn(x)
+        x = self.relu(x)
+        return x
+
+
+class QuantizableInception(googlenet_module.Inception):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(BasicConv2d=QuantizableBasicConv2d, *args, **kwargs)
+        self.cat = nn.quantized.FloatFunctional()
+
+    def forward(self, x):
+        outputs = self._forward(x)
+        return self.cat.cat(outputs, 1)
+
+
+class QuantizableGoogLeNet(googlenet_module.GoogLeNet):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(
+            BasicConv2d=QuantizableBasicConv2d,
+            Inception=QuantizableInception,
+            *args,
+            **kwargs
+        )
+        self.quant = torch.quantization.QuantStub()
+        self.dequant = torch.quantization.DeQuantStub()
+
+    def forward(self, x):
+        x = self._transform_input(x)
+        x = self.quant(x)
+        # N x 3 x 224 x 224
+        x = self.conv1(x)
```

Comment:
Can the part between the quant and dequant be shared with the float model by having a separate function?

Reply:
It can't because James added the following jit code after "x = self.fc(x)" to make float model scriptable:
`        if torch.jit.is_scripting():
            if not aux_defined:
                warnings.warn("Scripted GoogleNet always returns GoogleNetOutputs Tuple")
            return GoogLeNetOutputs(x, aux2, aux1)
        else:
            return self.eager_outputs(x, aux2, aux1)
`
In quantized model we need to insert "x = self.dequant(x)" between "x = self.fc(x)" and above jit code. The part between the quant and dequant can be shared if we don't need to make the model scriptable.

Reply:
I see, just want to make sure I understand the issue here. Is the code here not scriptable?

```python
def forward(self, x):
 x = self._transform_input(x)
 x = self.quant(x)
 aux_defined = self.training and self.aux_logits
 x, aux1, aux2 = self._common_forward(x)
 x = self.dequant(x)
 if torch.jit.is_scripting():
       if not aux_defined:
           warnings.warn("Scripted QuantizableGoogleNet always returns QuantizableGoogleNetOutputs Tuple")
            return QuantizableGoogLeNetOutputs(x, aux2, aux1)
      else:
            return self.eager_outputs(x, aux2, aux1)
```

Reply:
This didn't work with error:
>     x, aux1, aux2 = self. _common_forward(x)
> ValueError: not enough values to unpack (expected 3, got 1)

Reply:
Let's just rewrite the whole forward, this is fine. No need to make the base model more complicated